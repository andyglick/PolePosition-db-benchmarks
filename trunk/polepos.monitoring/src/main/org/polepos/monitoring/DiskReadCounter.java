/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import org.hyperic.sigar.DiskUsage;
import org.hyperic.sigar.SigarException;
import org.hyperic.sigar.SigarProxy;
import org.polepos.util.NoArgFunction;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.polepos.util.JavaLangUtils.rethrow;

/**
 * @author roman.stoffel@gamlor.info
 * @since 16.09.11
 */
public final class DiskReadCounter implements Sampler{
    final static String SAMPLER_NAME = "Disk-Reads on ";
    private final String machineName;
    private final List<DiskInfo> allDisks = new ArrayList<DiskInfo>();

    DiskReadCounter(Map<String,NoArgFunction<DiskUsage>> disks){
        this(disks,"client-machine");
    }
    DiskReadCounter(Map<String,NoArgFunction<DiskUsage>> disks,String machineName) {
        this.machineName = machineName;
        for (Map.Entry<String, NoArgFunction<DiskUsage>> entry : disks.entrySet()) {
            this.allDisks.add(new DiskInfo(entry.getKey(), entry.getValue()));
        }
    }

    public static Sampler create(SigarProxy sigar,String machineName) {
        try {
            Map<String,NoArgFunction<DiskUsage>> disks = new HashMap<String, NoArgFunction<DiskUsage>>();
            for (Object dn : sigar.getFileSystemMap().keySet()) {
                String diskName = (String)dn;

                final NoArgFunction<DiskUsage> disk = tryGetDiskOrReturnNull(sigar, diskName);
                if(null!=disk)  {
                    disks.put(diskName, disk);
                }
            }
            return new DiskReadCounter(disks,machineName);
        } catch (SigarException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    private static NoArgFunction<DiskUsage> tryGetDiskOrReturnNull(final SigarProxy sigar,final String diskName) {

        try {
            sigar.getDiskUsage(diskName);
        } catch (SigarException e) {
            return null;
        }
            return new NoArgFunction<DiskUsage>() {
                @Override
                public DiskUsage invoke() {
                    try {
                        return sigar.getDiskUsage(diskName);
                    } catch (SigarException e) {
                        throw rethrow(e);
                    }
                }
            };
    }

    @Override
    public MonitoringResult collectResult() {
        final DiskInfo currentRead = currentMostReadDisk();
        return MonitoringResult.create(MonitoringType.create(SAMPLER_NAME+currentRead.getName()+ machineName,"reads","kb"), (double)currentRead.currentRead() );
    }

    private DiskInfo currentMostReadDisk() {
        DiskInfo mostReads = allDisks.get(0);
        for (DiskInfo disk : allDisks) {
            if(disk.currentRead()>mostReads.currentRead()){
                mostReads = disk;
            }
        }
        return mostReads;
    }

    private static class DiskInfo{
        private final long initialRead;
        private final String name;
        private final NoArgFunction<DiskUsage> disk;

        public DiskInfo(String name,NoArgFunction<DiskUsage> disk) {
            this.name = name;
            this.disk = disk;
            initialRead = disk.invoke().getReads();
        }



        public long currentRead() {
            return disk.invoke().getReads()-initialRead;
        }

        public String getName() {
            return name;
        }
    }
}
