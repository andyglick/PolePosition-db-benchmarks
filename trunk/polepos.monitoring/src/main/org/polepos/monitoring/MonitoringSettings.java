/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import org.polepos.framework.PropertiesHandler;

import java.util.Arrays;
import java.util.Collection;

import static java.util.Collections.unmodifiableCollection;

/**
 * @author roman.stoffel@gamlor.info
 * @since 20.09.11
 */
public final class MonitoringSettings {
    static final String MONITORING_IS_ENABLED = "monitoring";
    static final String SAMPLERS = "samplers";
    public static final String REMOTE = "remote.collect";
    private static final String CONNECTOR = "agent.connector";
    private final boolean isEnabled;
    private final Collection<String> samplers;
    private final String remote;
    private final String connectionUrl;

    public MonitoringSettings(boolean enabled,String remote,String connectionUrl,String[] samplers) {
        isEnabled = enabled;
        this.remote = remote;
        this.connectionUrl = connectionUrl;
        this.samplers = Arrays.asList(samplers);
    }

    public static MonitoringSettings create(PropertiesHandler properties){
        return new MonitoringSettings(properties.getBoolean(MONITORING_IS_ENABLED),
                properties.get(REMOTE),
                properties.get(CONNECTOR),
                properties.getArray(SAMPLERS));
    }
    public static MonitoringSettings readFromConfig() {
        final PropertiesHandler properties = new PropertiesHandler(Monitoring.SETTINGS_FILE_NAME);
        return create(properties);
    }
    public String getConnectionUrl(){
        return connectionUrl;
    }

    public boolean isEnabled(){
        return isEnabled;
    }

    public Collection<String> getSamplers() {
        return unmodifiableCollection(samplers);
    }

    public String getRemote() {
        return remote;
    }
}
