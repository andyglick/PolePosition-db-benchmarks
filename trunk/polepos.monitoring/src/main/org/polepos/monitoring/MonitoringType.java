/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

/**
 * @author roman.stoffel@gamlor.info
 * @since 19.09.11
 */
public class MonitoringType {

    private final String name;
    private final String label;
    private final String unitOfMeasurment;

    private MonitoringType(String name, String label,String unitOfMeasurment) {
        this.name = name;
        this.label = label;
        this.unitOfMeasurment = unitOfMeasurment;
    }

    public String getName() {
        return name;
    }
    public String getLabel() {
        return label;
    }

    public String getUnitOfMeasurment() {
        return unitOfMeasurment;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MonitoringType that = (MonitoringType) o;

        if (!name.equals(that.name)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }

    @Override
    public String toString() {
        return "MonitoringType{" +
                "name='" + name + '\'' +
                '}';
    }
    public static MonitoringType create(String name) {
        return new MonitoringType(name,"","");
    }
    public static MonitoringType create(String name, String label,String unitOfMeasurment) {
        return new MonitoringType(name,label,unitOfMeasurment);
    }
//    public static String machineAppendix() {
//        try {
//            return " on "+ InetAddress.getLocalHost().getHostName();
//        } catch (UnknownHostException e) {
//            throw rethrow(e);
//        }
//    }
}
