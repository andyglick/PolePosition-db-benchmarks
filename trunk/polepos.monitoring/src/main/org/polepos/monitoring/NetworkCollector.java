/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import org.hyperic.sigar.NetInterfaceStat;
import org.hyperic.sigar.SigarException;
import org.hyperic.sigar.SigarProxy;

import java.util.ArrayList;
import java.util.List;

import static org.polepos.util.JavaLangUtils.rethrow;

/**
 * @author roman.stoffel@gamlor.info
 * @since 19.09.11
 */
public abstract class NetworkCollector implements Sampler {
    private final SigarProxy sigar;
    private final long initialValue;
    final static long KILO_BYTE = 1024;
    private final List<String> listOfDevices;
    private final MonitoringType type;

    protected NetworkCollector(MonitoringType type, SigarProxy sigar, List<String> listOfDevices) {
        this.sigar = sigar;
        this.initialValue = collectValue(sigar, listOfDevices);
        this.listOfDevices = listOfDevices;
        this.type = type;
    }

    @Override
    public MonitoringResult collectResult() {
        long currentValue = collectValue(sigar, listOfDevices);
        return MonitoringResult.create(type,(double) ((currentValue - initialValue) / KILO_BYTE));
    }

    public static Sampler createReceiveCollector(SigarProxy sigar) {
        List<String> listOfDevices = listOfMonitoredDevices(sigar);
        return new ReceiveNetworkCollector(sigar, listOfDevices);
    }
    public static Sampler createSendCollector(SigarProxy sigar) {
        List<String> listOfDevices = listOfMonitoredDevices(sigar);
        return new SendNetworkCollector(sigar, listOfDevices);
    }

    private static List<String> listOfMonitoredDevices(SigarProxy sigar)  {
        try {
            return filterOurUnusedDevices(sigar);
        } catch (SigarException e) {
            throw rethrow(e);
        }
    }

    private static List<String> filterOurUnusedDevices(SigarProxy sigar) throws SigarException {
        List<String> deviceNames = new ArrayList<String>();
        for (String device : sigar.getNetInterfaceList()) {
            final NetInterfaceStat starts = sigar.getNetInterfaceStat(device);
            if(inUse(starts)){
                deviceNames.add(device);
            }
        }
        return deviceNames;
    }

    private static boolean inUse(NetInterfaceStat starts) {
        return starts.getRxBytes()!=0||starts.getTxBytes()!=0;
    }

    private long collectValue(SigarProxy sigar, List<String> listOfDevices) {
        long value = 0;
        try {
            for (String ni : listOfDevices) {
                final NetInterfaceStat netInterfaceStat = sigar.getNetInterfaceStat(ni);
                final long rxBytes = readData(netInterfaceStat);
                value += rxBytes;
            }
        } catch (SigarException e) {
            throw rethrow(e);
        }
        return value;
    }
    protected abstract long readData(NetInterfaceStat netInterfaceStat);

    static class ReceiveNetworkCollector extends NetworkCollector{

        protected ReceiveNetworkCollector(SigarProxy sigar, List<String> listOfDevices) {
            super(MonitoringType.create("Network received bytes"+MonitoringType.machineNameAppendix(), "kb", "kbyte"),
                    sigar, listOfDevices);
        }

        protected long readData(NetInterfaceStat netInterfaceStat) {
            return netInterfaceStat.getRxBytes();
        }

    }

    static class SendNetworkCollector extends NetworkCollector{

        protected SendNetworkCollector(SigarProxy sigar, List<String> listOfDevices) {
            super(MonitoringType.create("Network sent bytes"+MonitoringType.machineNameAppendix(), "kb", "kbyte"),
                    sigar, listOfDevices);
        }

        protected long readData(NetInterfaceStat netInterfaceStat) {
            return netInterfaceStat.getTxBytes();
        }

    }
}
