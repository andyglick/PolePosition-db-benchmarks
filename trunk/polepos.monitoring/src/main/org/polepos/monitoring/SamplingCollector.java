/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import java.util.*;

/**
 * @author roman.stoffel@gamlor.info
 * @since 14.09.11
 */
final class SamplingCollector {
    private volatile List<Sampler> samplers;

    private SamplingCollector(Collection<? extends Sampler> samplers) {
        this.samplers = new ArrayList<Sampler>(samplers);
    }

    public static SamplingCollector start(Collection<? extends Sampler> samplers){
        return new SamplingCollector(samplers);
    }

    private List<MonitoringResult> sampleAndReturnResults() {
        ArrayList<MonitoringResult> results = new ArrayList<MonitoringResult>();
        for (Sampler sampler : samplers) {
            results.add(sampler.collectResult());
        }
        return results;
    }


    public Collection<MonitoringResult> stopAndCollectResults() {
        return sampleAndReturnResults();
    }
}
