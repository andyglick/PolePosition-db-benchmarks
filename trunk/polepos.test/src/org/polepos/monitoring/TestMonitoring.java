package org.polepos.monitoring;

import org.junit.Test;
import org.polepos.util.NoArgFunction;

import java.util.Collection;
import java.util.Collections;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static org.polepos.monitoring.MonitoringTestUtils.TEST_INTERVAL_IN_MILLISEC;
import static org.polepos.monitoring.MonitoringTestUtils.waitFor;

/**
 * @author roman.stoffel@gamlor.info
 * @since 14.09.11
 */
public class TestMonitoring {

    private static final int WAIT_TIME_IN_MILLISEC = 1000;

    @Test
    public void monitoringReturnsNotNull() {
        final LoadMonitoringResults results = Monitoring.monitor(testSettings(),
                Collections.<Sampler>emptyList(),new NoArgFunction<Object>() {
            @Override
            public Object invoke() {
                return null;
            }
        }).getMonitoring();
        assertNotNull(results);
    }

    private MonitoringSettings testSettings() {
        return new MonitoringSettings(true,TEST_INTERVAL_IN_MILLISEC);
    }

    @Test
    public void monitoringRunsSamplers() {
        CallCountingSampler sampler = new CallCountingSampler();
        final String ourResult = "result";
        final Monitoring.ResultAndData<String> results = Monitoring.monitor(testSettings(),Collections.singleton(sampler),
                new NoArgFunction<String> ()
                {
                    @Override
                    public String invoke() {
                        waitFor(WAIT_TIME_IN_MILLISEC);
                        return ourResult;
                    }
                });
        assertNotNull(results);
        assertEquals(ourResult, results.getData());
        sampler.assertHasBeenCalled();
    }

    @Test
    public void defaultSetIsNotNull() {
        final Collection<? extends Sampler> defaults = Monitoring.createDefaultMonitors();
        assertNotNull(defaults);
    }
    @Test
    public void canDisableMonitoring() {
        final MonitoringSettings settings = new MonitoringSettings(false, TEST_INTERVAL_IN_MILLISEC);
        final Collection<? extends Sampler> defaults = Monitoring.createDefaultMonitors();
        final LoadMonitoringResults results = Monitoring.monitor(settings, defaults, new NoArgFunction<Object>() {
            @Override
            public Object invoke() {
                waitFor(WAIT_TIME_IN_MILLISEC);
                return null;
            }
        }).getMonitoring();
        assertFalse(results.iterator().hasNext());
    }

    @Test
    public void returnsOriginalValue() {
        final Monitoring.ResultAndData results = Monitoring.monitor(testSettings(),Monitoring.createDefaultMonitors(),new NoArgFunction<String>() {
            @Override
            public String invoke() {
                return "hello world";
            }
        });

        assertEquals("hello world", results.getData());
        assertNotNull(results.getMonitoring());
    }
}
