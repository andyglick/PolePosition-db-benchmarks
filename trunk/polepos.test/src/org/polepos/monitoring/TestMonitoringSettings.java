/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import org.junit.Test;
import org.polepos.framework.PropertiesHandler;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * @author roman.stoffel@gamlor.info
 * @since 20.09.11
 */
public class TestMonitoringSettings {

    @Test
    public void isEnabled(){
        PropertiesHandler handler = monitoringCfg(true);
        MonitoringSettings toTest = MonitoringSettings.create(handler);
        assertTrue(toTest.isEnabled());
    }
    @Test
    public void isDisabled(){
        PropertiesHandler handler = monitoringCfg(false);
        MonitoringSettings toTest = MonitoringSettings.create(handler);
        assertFalse(toTest.isEnabled());
    }
    @Test
    public void samplingRate(){
        PropertiesHandler handler = samplingRate("100");
        MonitoringSettings toTest = MonitoringSettings.create(handler);
        assertEquals(100L,toTest.getSamplingRateInMillisec());
    }
    @Test
    public void defaultSampling(){
        PropertiesHandler handler = samplingRate(null);
        MonitoringSettings toTest = MonitoringSettings.create(handler);
        assertEquals(BackgroundSampling.INTERVAL_IN_MILLISEC,toTest.getSamplingRateInMillisec());
    }

    private PropertiesHandler monitoringCfg(boolean isMonitoringEnabled) {
        PropertiesHandler handler = mock(PropertiesHandler.class);
        when(handler.getBoolean(MonitoringSettings.MONITORING_IS_ENABLED)).thenReturn(isMonitoringEnabled);
        return handler;
    }
    private PropertiesHandler samplingRate(String samplingRate) {
        PropertiesHandler handler = mock(PropertiesHandler.class);
        when(handler.get(MonitoringSettings.MONITORING_SAMPLING_RATE)).thenReturn(samplingRate);
        return handler;
    }
}
