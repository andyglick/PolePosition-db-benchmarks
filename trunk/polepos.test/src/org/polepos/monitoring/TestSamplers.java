/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring;

import junit.framework.Assert;
import org.junit.Test;

import java.util.Collection;

/**
 * @author roman.stoffel@gamlor.info
 * @since 16.09.11
 */
public class TestSamplers {

    @Test
    public void returnsCPULoadMeter(){
        final Sampler sampler = Samplers.defaultInstance().cpuLoad();
        Assert.assertFalse(sampler.sample().getValue() ==0.0);
    }
    @Test
    public void returnDiskReadCounter(){
        final Sampler sampler = Samplers.defaultInstance().diskReads();
        Assert.assertTrue(sampler.sample().getValue() >= 0.0);
    }
    @Test
    public void returnNetwork(){
        final Sampler sampler = Samplers.defaultInstance().networkReads();
        Assert.assertTrue(sampler.sample().getValue() >= 0.0);
    }
    @Test
    public void doesContainAllSamplers(){
        final Collection<? extends Sampler> samplers = Samplers.defaultInstance().allSamplers();
        assertContains(CPULoadCollector.class,samplers);
        assertContains(NetworkReceiveCollector.class,samplers);
    }

    private void assertContains(Class<?> requiredType, Collection<? extends Sampler> samplers) {
        for (Sampler sampler : samplers) {
            if(requiredType.isInstance(sampler)){
                return;
            }
        }
        Assert.fail("Couldn't find "+requiredType+" in "+ samplers);
    }
}
