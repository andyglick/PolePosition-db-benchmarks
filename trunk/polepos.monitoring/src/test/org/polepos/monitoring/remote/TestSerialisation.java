/*
 * This file is part of the PolePosition database benchmark
 * http://www.polepos.org
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA  02111-1307, USA.MA  02111-1307, USA.
 */

package org.polepos.monitoring.remote;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;
import org.polepos.monitoring.LoadMonitoringResults;
import org.polepos.monitoring.MonitoringResult;
import org.polepos.monitoring.MonitoringType;

import java.util.Collections;

import static java.util.Arrays.asList;

/**
 * @author roman.stoffel@gamlor.info
 * @since 27.09.11
 */
public class TestSerialisation {

    private static final double DELTA = 0.0001;

    @Test
    public void canSerializeEmptyResult(){
        LoadMonitoringResults original = LoadMonitoringResults.create(Collections.<MonitoringResult>emptyList());

        LoadMonitoringResults serialized = Serialisation.fromJSON(Serialisation.toJSON(original));

        Assert.assertFalse(serialized.iterator().hasNext());
    }
    @Test
    public void emptyResultSerialized() throws Exception {
        LoadMonitoringResults original = LoadMonitoringResults.create(Collections.<MonitoringResult>emptyList());

        String serialized = Serialisation.toJSON(original);

        Approvals.approve(serialized);
    }
    @Test
    public void canSerializeResult(){
        MonitoringResult r1 = MonitoringResult.create(MonitoringType.percentUnit("result-1","r-1"),55 );
        MonitoringResult r2 = MonitoringResult.create(MonitoringType.create("result-2","r-2","runits",42),42 );
        LoadMonitoringResults original
                = LoadMonitoringResults.create(asList(r1,r2));

        LoadMonitoringResults serialized = Serialisation.fromJSON(Serialisation.toJSON(original));

        final MonitoringResult sr1 = serialized.byType(r1.getType());
        final MonitoringResult sr2 = serialized.byType(r2.getType());
        Assert.assertEquals(55.0,sr1.getValue(), DELTA);
        Assert.assertEquals(42.0,sr2.getValue(),DELTA);
        Assert.assertEquals("runits",sr2.getType().getUnitOfMeasurment());
    }
    @Test
    public void serializedForm() throws Exception {
        MonitoringResult r1 = MonitoringResult.create(MonitoringType.percentUnit("result-1","r-1"),55 );
        MonitoringResult r2 = MonitoringResult.create(MonitoringType.create("result-2","r-2","runits",42),42 );
        LoadMonitoringResults original
                = LoadMonitoringResults.create(asList(r1,r2));

        String serialized = Serialisation.toJSON(original);

        Approvals.approve(serialized);
    }
}
