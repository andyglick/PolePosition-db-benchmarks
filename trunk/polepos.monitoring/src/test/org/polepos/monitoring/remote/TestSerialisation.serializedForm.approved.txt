{"loadresult":{"results":[{"entry":[{"resulttype":{"name":"result-1","label":"r-1","unitOfMeasurment":"r"},"result":{"name":{"name":"result-1","label":"r-1","unitOfMeasurment":"r"},"value":55}},{"resulttype":{"name":"result-2","label":"r-2","unitOfMeasurment":"runits"},"result":{"name":{"name":"result-2","label":"r-2","unitOfMeasurment":"runits"},"value":42}}]}]}}